File: 71-simplify-path/simplify-path.java
class Solution {
    public String simplifyPath(String path) {
        int n = path.length();
        StringBuilder res = new StringBuilder();
        Stack<String> st = new Stack<>();
        
        for (int i = 0; i < n; i++) {
            if (path.charAt(i) == '/') continue;
            
            StringBuilder temp = new StringBuilder();
            while (i < n && path.charAt(i) != '/') {
                temp.append(path.charAt(i));
                i++;
            }
            
            String dir = temp.toString();
            if (dir.equals(".")) continue;
            else if (dir.equals("..")) {
                if (!st.isEmpty()) st.pop();
            } else {
                st.push(dir);
            }
        }
        
        while (!st.isEmpty()) {
            res.insert(0, st.pop());
            res.insert(0, "/");
        }
        
        if (res.length() == 0) {
            return "/";
        }
        
        return res.toString();
    }
}

File: 71-simplify-path/simplify-path.cpp
class Solution {
public:
    string simplifyPath(string path) {
        int n=path.size();
        string res="";
        stack<string> st;
        for(int i=0;i<n;i++){
            if(path[i]=='/') continue;
            string temp="";
            while(i<n && path[i]!='/'){
                temp+=path[i];
                i++;
            }
            if(temp==".") continue;
            else if(temp==".."){
                if(!st.empty()) st.pop();
            }
            else st.push(temp);
        }
        while(!st.empty())
        {
            res = "/" + st.top() + res;
            st.pop();
        }
        if(res.size() == 0)
            return "/";
        
        return res;
    }
};